### PR TITLE
Introduce Function.constLazy & constMemoized

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: src/library/scala/Function.scala [hunk 1]
```diff
@@ -27,6 +27,14 @@ object Function {
   /** The constant function */
   def const[T, U](x: T)(y: U): T = x
 
+  def constLazy[T, U](x: => T)(y: U): T = x
```
Comment (NthPortal) [inline src/library/scala/Function.scala:30]: docs would be nice, particularly for this one which one might think should only evaluate `x` once (I realise that's what `constMemoized` does). I think this is probably the best name there is for it though.

**Comment by NthPortal** [comment]
this needs to be 2.14.x

File: src/library/scala/Function.scala (hunk 1 above)
Comment (dwijnand) [reply src/library/scala/Function.scala:30]: Yeah, the naming convention I've seen is "lazy" for "by-name".  It doesn't help that the `lazy` keyword is only applicable to `val`, to make it "by-need" or memoized laziness.  I could've gone with `constByName` and `constByNeed` but then I'm not used to those names enough to remember which is which on the fly.

I'll doc it up.

**Comment by dwijnand** [comment]
> the naming convention I've seen is "lazy" for "by-name".

Hmm, reading https://en.wikipedia.org/wiki/Lazy_evaluation it seems "lazy" implies memoized.  So the naming convention used here might be a terrible idea... 🤔

**Comment by joroKr21** [comment]
In Scala lazy means by-need. The default is by-value and by-name is `=> T`.

**Comment by dwijnand** [comment]
Yeah, that's what I'm realising.  What do you think we should call the lazy, memoizing constant function, and what should we call the non-lazy, non-memoizing constant function?

**Comment by sjrd** [comment]
> What do you think we should call the lazy, memoizing constant function

This is lazy. Call it `constLazy`.

> and what should we call the non-lazy, non-memoizing constant function?

Don't define it at all. It's shorter to write it down at call site: `Function.theThing(x)` is equivalent to `_ => x`.

**Comment by dwijnand** [comment]
I care to define it because it helps parameter type inference:

```scala
Welcome to Scala 2.13.0 (OpenJDK GraalVM CE 19.0.0, Java 1.8.0_212).
Type in expressions for evaluation. Or try :help.

scala> val pf: PartialFunction[String, String] = { case "hi" => "bye" }
pf: PartialFunction[String,String] = <function1>

scala> pf.applyOrElse("x", _ => "k")
                           ^
       error: missing parameter type

scala> def theThing[T, U](x: => T)(y: U): T = x
constLazy: [T, U](x: => T)(y: U)T

scala> pf.applyOrElse("x", theThing("k"))
res1: String = k
```

**Comment by dwijnand** [comment]
Naming paralysis

**Pull request closed without merging**